File: crates/codegraph-scala/tests/fixtures/sample_app.scala
import scala.collection.mutable
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global

abstract class Entity(val id: Int, val name: String) {
  def display(): String
}

class User(id: Int, name: String, val email: String) extends Entity(id, name) {
  private val roles = mutable.ListBuffer[String]()

  override def display(): String = s"User($name, $email)"

  def addRole(role: String): Unit = {
    if (!roles.contains(role)) {
      roles += role
      validateRole(role)
    }
  }

  def isAdmin: Boolean = roles.contains("admin")

  private def validateRole(role: String): Unit = {
    val allowed = Set("admin", "user", "guest")
    if (!allowed.contains(role)) {
      throw new IllegalArgumentException(s"Invalid role: $role")
    }
  }
}

class Product(id: Int, name: String, val price: Double) extends Entity(id, name) {
  override def display(): String = s"Product($name, $$$price)"

  def discountedPrice(percent: Double): Double = {
    calculateDiscount(price, percent)
  }

  private def calculateDiscount(original: Double, percent: Double): Double = {
    original * (1 - percent / 100)
  }
}

object UserService {
  private val users = mutable.ListBuffer[User]()

  def createUser(name: String, email: String): Future[User] = Future {
    val user = new User(users.length + 1, name, email)
    users += user
    sendWelcomeEmail(user)
    user
  }

  def findUser(id: Int): Option[User] = {
    users.find(_.id == id)
  }

  private def sendWelcomeEmail(user: User): Unit = {
    println(s"Welcome ${user.name}!")
  }
}

trait Serializable {
  def toJson: String
}

File: crates/codegraph-dart/tests/fixtures/sample_app.dart
import 'dart:async';
import 'dart:convert';

abstract class Entity {
  final int id;
  final String name;
  Entity(this.id, this.name);
  String display();
}

class User extends Entity {
  final String email;
  List<String> _roles = [];

  User(int id, String name, this.email) : super(id, name);

  @override
  String display() => 'User($name, $email)';

  void addRole(String role) {
    if (_roles.contains(role)) return;
    _roles.add(role);
    _validateRole(role);
  }

  bool get isAdmin => _roles.contains('admin');

  void _validateRole(String role) {
    final allowed = ['admin', 'user', 'guest'];
    if (!allowed.contains(role)) {
      throw ArgumentError('Invalid role: $role');
    }
  }
}

class Product extends Entity {
  final double price;

  Product(int id, String name, this.price) : super(id, name);

  @override
  String display() => 'Product($name, \$$price)';

  double discountedPrice(double percent) {
    return _calculateDiscount(price, percent);
  }

  double _calculateDiscount(double original, double percent) {
    return original * (1 - percent / 100);
  }
}

class UserService {
  final List<User> _users = [];

  Future<User> createUser(String name, String email) async {
    final user = User(_users.length + 1, name, email);
    _users.add(user);
    await _sendWelcomeEmail(user);
    return user;
  }

  User? findUser(int id) {
    return _users.where((u) => u.id == id).firstOrNull;
  }

  Future<void> _sendWelcomeEmail(User user) async {
    print('Welcome ${user.name}!');
  }
}
